source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst

```diff
@@ -1088,57 +1089,59 @@ updateRSWithPrimaryFromMember
 
-    if description.address not in topologyDescription.servers:
+    if serverDescription.address not in topologyDescription.servers:
         return
 
     if topologyDescription.setName is null:
-        topologyDescription.setName = description.setName
+        topologyDescription.setName = serverDescription.setName
 
-    else if topologyDescription.setName != description.setName:
+    else if topologyDescription.setName != serverDescription.setName:
         # We found a primary but it doesn't have the setName
         # provided by the user or previously discovered.
         remove this server from topologyDescription and stop monitoring it
         checkIfHasPrimary()
         return
 
-    if description.setVersion is not null and description.electionId is not null:
-        # Election ids are ObjectIds, see
-        # "using setVersion and electionId to detect stale primaries"
-        # for comparison rules.
-        if (topologyDescription.maxSetVersion is not null and
-            topologyDescription.maxElectionId is not null and (
-                topologyDescription.maxSetVersion > description.setVersion or (
-                    topologyDescription.maxSetVersion == description.setVersion and
-                    topologyDescription.maxElectionId > description.electionId
-                )
-            ):
-
-            # Stale primary.
-            replace description with a default ServerDescription of type "Unknown"
-            checkIfHasPrimary()
-            return
-
-        topologyDescription.maxElectionId = description.electionId
-
-    if (description.setVersion is not null and
-        (topologyDescription.maxSetVersion is null or
-            description.setVersion > topologyDescription.maxSetVersion)):
+    # Election ids are ObjectIds, see
+    # "using setVersion and electionId to detect stale primaries"
+    # for comparison rules.
+
+    # Null values for both electionId and setVersion are always considered less than
+    if topologyDescription.maxElectionId > serverDescription.electionId or (
+        (
+            topologyDescription.maxElectionId == serverDescription.electionId
+            or serverDescription.electionId == null
+        )
+        and topologyDescription.maxSetVersion > serverDescription.setVersion
+    ):
+        # Stale primary.
+        # replace serverDescription with a default ServerDescription of type "Unknown"
+        checkIfHasPrimary()
+        return
 
-        topologyDescription.maxSetVersion = description.setVersion
+    if topologyDescription.maxElectionId < serverDescription.electionId or (
+        (
+            topologyDescription.maxElectionId == serverDescription.electionId
+            or serverDescription.electionId == null
+        )
+        and topologyDescription.maxSetVersion < serverDescription.setVersion
+    ):
+        topologyDescription.maxElectionId = serverDescription.electionId
+        topologyDescription.maxSetVersion = serverDescription.setVersion
```

Comment:
I would prefer to see this if-stmt and the "# Stale primary" be combined into one if/else for simplicity. If we are leaving the null comparison implied and not explicit (which the current code does) then what do you think about this?:
```
if (topologyDescription.electionId == serverDescription.maxElectionId and topologyDescription.setVersion >= serverDescription.maxSetVersion) or topologyDescription.electionId > serverDescription.maxElectionId:
    topologyDescription.maxElectionId = serverDescription.electionId
    topologyDescription.maxSetVersion = serverDescription.setVersion
else:
    # Stale primary....
```

Reply:
Went with just the first way for now, We could maybe expand the null comment with the suggestions for zero OID/setVersion, I'm thinking its too verbose for pseudo code.